### PR TITLE
Backports from Parabola

### PR DESCRIPTION
**Pull request by LukeShu**
These are commits from Parabola's dbscripts that I think are
upstreamable.  No real theme to them, but because some of them are
wide-sweeping (eg. clean up quoting), they are a good base layer to
apply before backporting other changes.

I've done my best to look over them, but at least one weird thing did
happen while rebasing them, so it's possible I missed something that
rebased weird.

Some of this duplicates (non-backported) changes in the last patchset
I submitted.

This is currently v2 of the patchset.  v1 was submitted on the mailing list.  I'm submitting v2 here to get a shiny little confirmation from Travis that I didn't break anything; I'm submitting it to the mailing list as well.

### CODE REVIEW

**Comment by LukeShu** [comment]
Huh, Travis ran slightly faster than my local `make test`.  Eli's awk expression in the final commit.  I've pushed a revised version.  I'm going to bed.  If in the morning Travis/`make test` came back saying we're good, I'll also send it to the mailing list.

File: cron-jobs/sourceballs [hunk 1]
```diff
@@ -53,22 +53,16 @@ for f in "${FTP_BASE}"/${SRCPOOL}/*${SRCEXT}; do
...
-		pkglicense=(${pkginfo[@]:3})
-
+	while read -r pkgbase pkgver pkgarch pkglicense; do
 		# Should this package be skipped?
 		if grep -Fqx "${pkgbase}" "${dirname}/sourceballs.skip"; then
 			continue
 		fi
 		# Check if the license or .force file does not enforce creating a source package
-		if ! ([[ -z ${ALLOWED_LICENSES[@]} ]] || chk_license ${pkglicense[@]} || grep -Fqx "${pkgbase}" "${dirname}/sourceballs.force"); then
+		if ! ([[ ${#ALLOWED_LICENSES[@]} == 0 ]] || chk_license "${pkglicense[@]}" || grep -Fqx "${pkgbase}" "${dirname}/sourceballs.force"); then
```
Comment (eli-schwartz) [inline cron-jobs/sourceballs]: Instead of string comparisons why not use shell arithmetic?

File: db-functions [hunk 2]
```diff
@@ -282,7 +282,7 @@ getpkgfile() {
 
 getpkgfiles() {
 	local f files
-	if [[ ! -z "$(printf '%s\n' "${@%\.*}" | sort | uniq -D)" ]]; then
+	if ! printf '%s\n' "${@%\.*}" | awk 'a[$0]++>1{exit 1}'; then
```
Comment (eli-schwartz) [inline db-functions]: Why not fix this all at once instead of changing the same line multiple times?

File: db-functions [hunk 3]
```diff
@@ -303,11 +303,7 @@ check_pkgfile() {
 
 	in_array "${pkgarch}" "${ARCHES[@]}" 'any' || return 1
 
-	if echo "${pkgfile##*/}" | grep -q "^${pkgname}-${pkgver}-${pkgarch}"; then
-		return 0
-	else
-		return 1
-	fi
+	[[ ${pkgfile##*/} == "${pkgname}-${pkgver}-${pkgarch}"* ]]
```
Comment (eli-schwartz) [inline db-functions]: Don't use ==, use = instead. This is a weird bash extension that literally serves no purpose as it merely aliases =.

File: cron-jobs/check_archlinux/parse_pkgbuilds.sh [hunk 4]
```diff
@@ -6,23 +6,23 @@
...
 	done
 }
 
 restore_package_variables() {
-	for var in ${splitpkg_overrides[@]}; do
+	for var in "${splitpkg_overrides[@]}"; do
 		indirect="${var}_backup"
 		if [ -n "${!indirect}" ]; then
-			eval "${var}=(\${$indirect[@]})"
+			eval "${var}=(\"\${$indirect[@]}\")"
```
Comment (eli-schwartz) [inline cron-jobs/check_archlinux/parse_pkgbuilds.sh]: I would still like to fix this by avoiding the issue entirely and using `declare` as I mentioned on the mailing list.

File: test/cases/db-update.bats [hunk 5]
```diff
@@ -222,7 +222,7 @@ load ../lib/common
 
 @test "package has to be aregular file" {
 	local p
-	local target=$(mktemp -d)
+	local target=$(mktemp -dt)
```
Comment (eli-schwartz) [inline test/cases/db-update.bats]: No TEMPLATE given, this change is extraneous.

File: db-functions (hunk 2 above)
Comment (LukeShu) [reply db-functions]: It's two logically distinct changes.

File: cron-jobs/check_archlinux/parse_pkgbuilds.sh (hunk 4 above)
Comment (LukeShu) [reply cron-jobs/check_archlinux/parse_pkgbuilds.sh]: 1. From the ML:

> Maybe I should fix the backups as well, but that is a slightly more complicated case there [in makepkg].

Is it?  It looks like the exact same case there.

2. What about this?

```
backup_package_variables() {
	restore_package_variables=$(declare -p "${splitpkg_overrides[@]}" 2>/dev/null || true)
}

restore_package_variables() {
	unset "${splitpkg_overrides[@]}"
	eval "$restore_package_variables"
}
```

File: test/cases/db-update.bats (hunk 5 above)
Comment (LukeShu) [reply test/cases/db-update.bats]: Whoops, you are correct.

File: db-functions (hunk 2 above)
Comment (eli-schwartz) [reply db-functions]: And so are several other changes you lumped into the first commit.

**Comment by LukeShu** [comment]
> As I said on the mailing list, I don't see the need to overquote variables everywhere. You want it for the sake of linting (this is a shellcheck thing I actually disagree with) and to reduce the diff against your fork (which is not our design goal).

> I'm not sure what to do about it.

I respect that.  I'm not sure what I would do about it if I were in your place.  If you want me to re-submit this patchset without the quoting commit, let me know.

File: db-functions (hunk 3 above)
Comment (eli-schwartz) [reply db-functions]: No, it is a cool bash extension that serves a purpose. You've added the only two examples of using == in the main code, just now, and I dispute Chet Ramey's reasons for adding the == alias to the bash scripting language.

File: db-functions (hunk 2 above)
Comment (LukeShu) [reply db-functions]: They're all "this array wasn't quoted, but we can do a better fix than simply quoting it"

File: db-functions (hunk 3 above)
Comment (LukeShu) [reply db-functions]: Glob/regex matching, but the rest of it is weird.  `[[ -r $filename ]]` adds nothing to the language, and serves no purpose over `[ -r "$filename" ]`

At the end of the day, it's just a style choice.  And that's fine.  But "it's the codebase's style" is the winning argument there, not "it's a bad extension".  I hadn't realized that the main codebase was all `=`; I'd though it was inconsistent; I hadn't realized that all of the `[[ == ]]` uses were in `test/`.

**Comment by LukeShu** [comment]
I've just pushed v3 of the patchset:

v3:
 - Use `[[ $v = true ]]` for testing boolean variables
 - Split the initial quoting commit in to a "only add double-quotes"
   commit and a commit with more involved changes
 - Combine the `sort|uniq` -> `awk` replacement commit with the more
   involved quoting commit
 - `db-functions:check_pkgfile`: Use `[[ = ]]` instead of `[[ == ]]`
 - Don't bother with `mktemp`'s `-t` flag if there isn't a template given
 - Use `declare -p` to simplify `{backup,restore}_package_variables`
 - Use `(( ))` instead of `[[ ]]` when checking if an array is empty

I believe it addresses all of Eli's concerns, though it still quotes variables (but not variable-used-as-a-command; we don't have those anymore).  I would have never though that quoting variables would be a controversial change.

File: cron-jobs/check_archlinux/parse_pkgbuilds.sh [hunk 6]
```diff
@@ -10,21 +10,12 @@ variables=('pkgname' 'pkgbase' 'epoch' 'pkgver' 'pkgrel' 'makedepends' 'arch' ${
...
-	for var in ${splitpkg_overrides[@]}; do
-		indirect="${var}_backup"
-		if [ -n "${!indirect}" ]; then
-			eval "${var}=(\${$indirect[@]})"
-		else
-			unset ${var}
-		fi
-	done
+	unset "${splitpkg_overrides[@]}"
+	eval "$restore_package_variables"
```
Comment (eli-schwartz) [inline cron-jobs/check_archlinux/parse_pkgbuilds.sh]: FWIW: https://lists.archlinux.org/pipermail/pacman-dev/2018-March/022406.html

TBH this whole script seems somewhat convoluted. If anything I think I'd rather port `check_packages.py` to use pyalpm instead of ctypes, and `makepkg --printsrcinfo` + [python-srcinfo](https://github.com/kyrias/python-srcinfo) rather than `parse_pkgbuilds.sh`

But then again, I don't think anyone still uses it...

Comment (LukeShu) [reply cron-jobs/check_archlinux/parse_pkgbuilds.sh]: Agree about it being convoluted.

Isn't your solution going to have `declare -p` spit things on stderr?

File: test/lib/common.bash [hunk 7]
```diff
@@ -269,7 +269,9 @@ checkRemovedPackageDB() {
 		for tarch in ${tarches[@]}; do
 			if [ -r "${FTP_BASE}/${repo}/os/${tarch}/${repo}${db%.tar.*}" ]; then
 				for pkgname in ${pkgnames[@]}; do
-					echo "$(bsdtar -xf "${FTP_BASE}/${repo}/os/${tarch}/${repo}${db%.tar.*}" -O)" | grep -qv ${pkgname}
+					if bsdtar -xf "${FTP_BASE}/${repo}/os/${tarch}/${repo}${db%.tar.*}" -O | grep ${pkgname} &>/dev/null; then
+						return 1
+					fi
```
Comment (eli-schwartz) [inline test/lib/common.bash]: grep -v has been converted to if grep; then return 1

I understand why that has been done here, as AFAICT this used to always return successfully since there was always some line that did not contain ${pkgname} -- even if that line was `%NAME%` :D 

Which means we weren't actually testing anything with checkRemovedPackageDB as we cannot actually get test failures.

That being said, this does need to be noted in a commit message.

I also think it would make more sense to check for a pkgname using bsdtar -tf whereas in the other change we need to use -xf in order to check a pkgfile. (I'm not positive why we need to check the pkgfile there either.)

File: testing2x [hunk 8]
```diff
@@ -49,7 +49,7 @@ done
 for pkgarch in ${ARCHES[@]}; do
 	repo_unlock ${TESTING_REPO} ${pkgarch}
 done
-for repo in ${STABLE_REPOS[@]}; do
+for repo in  "${STABLE_REPOS[@]}"; do
```
Comment (eli-schwartz) [inline testing2x]: Added extra whitespace here, commit message says you are removing whitespace.

File: db-functions [hunk 9]
```diff
@@ -433,9 +431,9 @@ set_repo_permission() {
 
 	if [[ -w ${dbfile} ]]; then
 		local group=$(/usr/bin/stat --printf='%G' "$(dirname "${dbfile}")")
-		chgrp $group "${dbfile}" || error "Could not change group of %s to %s" "$dbfile" "$group"
+		chgrp "$group" "${dbfile}"    || error "Could not change group of %s to %s" "$dbfile" "$group"
 		chgrp $group "${filesfile}" || error "Could not change group of %s to %s" "$filesfile" "$group"
-		chmod g+w "${dbfile}" || error "Could not set write permission for group %s to %s" "$group" "$dbfile"
+		chmod g+w "${dbfile}"    || error "Could not set write permission for group %s to %s" "$group" "$dbfile"
 		chmod g+w "${filesfile}" || error "Could not set write permission for group %s to %s" "$group" "$filesfile"
```
Comment (eli-schwartz) [inline db-functions]: You lined up one of these, but the other is still misaligned...

File: cron-jobs/sourceballs [hunk 10]
```diff
@@ -56,19 +56,13 @@ done | sort -u > "${WORKDIR}/available-src-pkgs"
...
 	newpkgs=()
 	failedpkgs=()
-	while read line; do
-		pkginfo=(${line})
-		pkgbase=${pkginfo[0]}
-		pkgver=${pkginfo[1]}
-		pkgarch=${pkginfo[2]}
-		pkglicense=(${pkginfo[@]:3})
-
+	while read -r pkgbase pkgver pkgarch pkglicense; do
```
Comment (eli-schwartz) [inline cron-jobs/sourceballs]: pkglicense is now a string, rather than an array. It would still work as both licenses are in the (unquoted) array element 0, but of course that is now quoted.

So this only works if there is exactly one license in play.

File: db-functions (hunk 9 above)
Comment (LukeShu) [reply db-functions]: Whoops.  That line is 2 characters to the left because `$group` is unquoted--it gets quoted in the next commit.  I missed that line as being part of lining up when I split it in to 2 commits.

File: testing2x (hunk 8 above)
Comment (LukeShu) [reply testing2x]: Huh.  I wonder how I screwed that up.  But I did remove whitespace further up in the commit!

**Comment by LukeShu** [comment]
I've just pushed v4 of the patchset

v4:
 - db-functions: fix lining up (relied on a change in a later commit)
 - testing2x: fix whitespace being introduced
 - parse_pkgbuilds.sh: have backup_package_variables output an
   expression; remove restore_package_variables
 - sourceballs: pkglicense should be an array
 - test/: have a test fixture have multiple entries in license=()
 - Note in the commit message that the `grep -q` commit fixes an issue
   by avoiding `grep -v`.

**Comment by LukeShu** [comment]
I've just pushed v5 of this patchset.  It's just rebasing on to
master.

v5:
 - Rewrite printf-formatter commit message, since most of this commit
   was subsumed by 33aae31.
 - Fix the `${pkgs[@]@Q}` mistake in #6 as part of the
   fixups-near-quoting commit.

**Comment by LukeShu** [comment]
(whoops, deleted the wrong branch)

**Comment by LukeShu** [comment]
v6:
 - rebase, shuffle things affecting `arch_repo_modify` between commits

`git diff` between v5 and v6 comes up empty.

**Comment by LukeShu** [comment]
v7:
 - I had missed `$_lockblock` being used as a boolean/command in db-functions

**Comment by eli-schwartz** [comment]
Was looking at replacing the whole parse_pkgbuild.sh mess with python-srcinfo and a custom srcinfo generator, but then got sidetracked and had to debug this: https://lists.archlinux.org/pipermail/pacman-dev/2018-March/022428.html

yay...

File: db-functions [hunk 11]
```diff
@@ -273,16 +273,16 @@ getpkgfile() {
...
 	fi
 
-	echo ${1}
+	echo "${1}"
 }
 
 getpkgfiles() {
 	local f files
-	if [[ ! -z "$(echo ${@%\.*} | sed "s/ /\n/g" | sort | uniq -D)" ]]; then
+	if ! printf '%s\n' "${@%\.*}" | awk 'a[$0]++>1{exit 1}'; then
```
Comment (eli-schwartz) [inline db-functions]: My initial suggestion did not have >1 in it, and this version does not work.

**Comment by LukeShu** [comment]
v8:
 - db-functions: getpkgfiles: fix awk expression to detect duplicates
 - test/: add a test to verify that getpkgfiles() rejects duplicates

**Comment by LukeShu** [comment]
v9:
 - rebase

I think you should at least pick the `test/` commit, but I wouldn't be upset if you wanted to reject the others.

**Comment by eli-schwartz** [comment]
(Apologies for my initial response to the quoting changes, I did sort of overreact when I saw it breaking things due to excessive quoting. But that appears to be solved now, and it doesn't look like any other regressions are hiding.)

The rest look okay, I think. But I'm slightly confused what the test/ commit does, it seems to overlap the test case right above it?

Also apologies for the lateness of the response, since I'm now on holiday and somewhat erratically available.

**Comment by LukeShu** [comment]
The test above it checks that we detect when a package in `$STAGING` duplicates a package already in the repos.  The new test checks that we detect duplicates when *both* packages are in `$STAGING`.

v10:
 - clean rebase, no changes

**Comment by eli-schwartz** [comment]
v11: remove veverything comments from commit messages
